// plugwise-stretch/stretch.js
"use strict";

var XML = require('pixl-xml');
var request = require('request');
const EventEmitter = require('events').EventEmitter;

class Stretch extends EventEmitter {

	constructor(password, ip, id, hostname) {
		super();

		// Store properties
		this.password = password;
		this.ip = ip;
		this.id = id;
		this.hostname = hostname;
		this.onoff = undefined;
		this.status = 2;

		// Start polling for data changes
		this._startPolling()
	}

	setState(state, callback) {

		var toggle = (state) ? "on" : "off";

		request({
			url: this._createAPIUrl() + '/relay',
			method: 'PUT',
			body: '<relay><state>' + toggle + '</state></relay>',
			headers: { 'Content-Type': 'text/xml' }
		}, function (error) {
			if (error) {

				console.log("Stretch: set state error: " + error);

				// Add one to status
				this.status++;

				// Emit device unavailable
				if (this.status >= 2) {
					this.emit("unavailable", this);
					this.status = 0;
				}
			}
			else {

				// Reset status
				this.status = 0;

				console.log("Stretch: switch state to " + toggle);

				// Update interval value
				this.onoff = state;

				// Event event that onoff state changed
				this.emit("onoff", state);

				// Return with callback
				callback(null, state);
			}
		}.bind(this));
	};

	_getState(callback) {
		request({ url: this._createAPIUrl(), timeout: 8000, method: 'GET' }, function (error, response, body) {
			if (error) {

				console.log("Stretch: get state error: " + error);

				// Add one to status
				this.status++;

				// Emit device unavailable
				if (this.status >= 2) {
					this.emit("unavailable", this);
					this.status = 0;
				}
			}
			else {

				// Fetch data from XML
				var doc = XML.parse(body);
				if (doc && doc.appliance
					&& doc.appliance.actuators
					&& doc.appliance.actuators.relay
					&& doc.appliance.actuators.relay.state) {

					// Emit device available
					if (this.status >= 2) {
						this.emit("available", this);
						this.status = 0;
					}

					// Reset status
					this.status = 0;

					// Convert to boolean
					let state = (doc.appliance.actuators.relay.state === "on");

					// If value changed
					if (typeof this.onoff != "undefined" && this.onoff != state) {

						// Emit onoff update event
						this.emit("onoff", state);
					}

					// Update internal value
					this.onoff = state;

					// Return result
					if (typeof callback == "function") callback(null, state);
				}
				else {

					console.log("Stretch: get state invalid data from plugwise api");

					// Add one to status
					this.status++;

					// Emit device unavailable
					if (this.status >= 2) {
						this.emit("unavailable", this);
						this.status = 0;
					}

					// Callback error
					if (typeof callback == "function") callback(true, false);
				}

				if (typeof this.onoff === "undefined") this.onoff = false;
			}
		}.bind(this));
	}

	_startPolling() {

		// Fill initial values
		this._getState();

		this.pollInterval = setInterval(() => {

			// Update values
			this._getState();

		}, 15000);
	}

	_createAPIUrl() {
		return `http://stretch:${this.password}@${this.ip}/core/appliances/${this.id}`;
	}

	remove() {
		if (this.pollInterval) clearInterval(this.pollInterval);
	}
}

module.exports = Stretch;
